src/frame-validator.cpp

```diff
@@ -3,90 +3,106 @@
 
 #include "frame-validator.h"
 
+#define INVALID_PIXELS_THRESHOLD 0.1
+
 namespace librealsense
 {
     void frame_validator::on_frame(rs2_frame * f)
     {
-        if (propagate((frame_interface*)f))
+        if (!_stopped && propagate((frame_interface*)f))
         {
-            _user_callback->on_frame(f);
-            _sensor->set_frames_callback(_user_callback);
+            if(is_user_requseted_frame((frame_interface*)f))
+                _user_callback->on_frame(f);
         }
     }
 
     void frame_validator::release()
-    {
-
-    }
+    {}
 
-    frame_validator::frame_validator(std::shared_ptr<sensor_interface> sensor, frame_callback_ptr user_callback, stream_profiles current_requests) :
+    frame_validator::frame_validator(std::shared_ptr<sensor_interface> sensor, frame_callback_ptr user_callback, stream_profiles current_requests, stream_profiles validator_requests) :
         _sensor(sensor), 
         _user_callback(user_callback), 
-        _current_requests(current_requests)
-    {
-    
-    }
+        _user_requests(current_requests),
+        _validator_requests(validator_requests),
+        _stopped(false),
+        _validated(false)
+    {}
 
     frame_validator::~frame_validator()
-    {
-
-    }
+    {}
 
     bool frame_validator::propagate(frame_interface* frame)
     {
-        auto vf = dynamic_cast<video_frame*>(frame);
-        if (vf == nullptr) {
+        if(_validated)
             return true;
-        }
+
+        auto vf = dynamic_cast<video_frame*>(frame);
+        if (vf == nullptr)
+            throw std::runtime_error(to_string() << "non video stream arrived to frame_validator");
+
         auto stream = vf->get_stream();
-        if (stream->get_stream_type() != RS2_STREAM_DEPTH)
+
+        //all streams except ir will be droped untill the validation on ir,
+        //after the validation all streams will be passeded to user callback directly
+        if (stream->get_stream_type() != RS2_STREAM_INFRARED)
+            return false;
```

Comment:
If for some reason Ir frames do not arrive for a considerable amount of time but others do - is it possible to generate a log, like after 100 consecutive "return false" ?